modules/replicator/src/main/java/org/apache/ignite/internal/replicator/Replica.java

```diff
@@ -295,15 +304,24 @@ private CompletableFuture<LeaseGrantedMessageResponse> processLeaseGrantedMessag
     private CompletableFuture<Void> processWaitReplicaStateMessage(WaitReplicaStateMessage msg) {
         LOG.info("WaitReplicaStateMessage was received [groupId = {}]", groupId());
 
-        return waitForActualState(FastTimestamps.coarseCurrentTimeMillis() + TimeUnit.SECONDS.toMillis(msg.timeout()));
+        // TODO: https://issues.apache.org/jira/browse/IGNITE-22122
+        return waitForActualState(FastTimestamps.coarseCurrentTimeMillis() + msg.timeout())
+                .thenComposeAsync(
+                        v -> sendPrimaryReplicaChangeToReplicationGroup(msg.enlistmentConsistencyToken()),
+                        executor
+                );
     }
 
     private CompletableFuture<Void> sendPrimaryReplicaChangeToReplicationGroup(long leaseStartTime) {
         PrimaryReplicaChangeCommand cmd = REPLICA_MESSAGES_FACTORY.primaryReplicaChangeCommand()
                 .leaseStartTime(leaseStartTime)
                 .build();
 
-        return raftClient.run(cmd);
+        return raftClient.run(cmd).thenRun(() -> {
+            if (!waitForActualStateFuture.compareAndSet(null, nullCompletedFuture())) {
```

Comment:
Why do we need this here?

Reply:
The methodoid is the final phase of the wating for the actual state process.